payments_service/braintree/views.py

```diff
@@ -1,6 +1,97 @@
 
 
 class TokenGenerator(SolitudeBodyguard):
+    """
+    Generate a client token to begin processing payments.
+    """
     methods = ['post']
     resource = 'braintree.token.generate'
+
+
+class Subscribe(APIView):
+    """
+    Subscribe the logged in buyer to a plan.
+    """
+    def post(self, request):
+        form = SubscriptionForm(request.DATA)
+        if not form.is_valid():
+            return error_400(response=form.errors)
+
+        buyer_uuid = request.user.pk
+        api = solitude.api()
+        try:
+            self.set_up_customer(buyer_uuid)
+            pay_method_uri = self.get_pay_method(
+                buyer_uuid,
+                form.cleaned_data['pay_method_uri'],
+                form.cleaned_data['pay_method_nonce'])
+
+            api.braintree.subscription.post({
+                'paymethod': pay_method_uri,
+                'plan': form.cleaned_data['plan_id'],
+            })
+        except HttpClientError, exc:
+            return error_400(exception=exc)
+
+        return Response({}, status=204)
+
+    def get_pay_method(self, buyer_uuid, pay_method_uri, pay_method_nonce):
+        api = solitude.api()
+        if not pay_method_uri:
+            log.info('creating new payment method for buyer {b}'
+                     .format(b=buyer_uuid))
+            pay_method = api.braintree.paymethod.post({
+                'buyer_uuid': buyer_uuid,
+                'nonce': pay_method_nonce,
+            })
+            pay_method_uri = pay_method['mozilla']['resource_uri']
+        else:
+            log.info('paying with saved payment method {m} for buyer {b}'
+                     .format(b=buyer_uuid, m=pay_method_uri))
+
+        return pay_method_uri
+
+    def set_up_customer(self, buyer_uuid):
+        api = solitude.api()
+        # TODO: This can be simplified after:
+        # https://github.com/mozilla/payments-service/issues/22
+        buyer_results = api.generic.buyer.get(uuid=buyer_uuid)
+        if len(buyer_results) == 1:
+            buyer = buyer_results[0]
+        else:
+            raise ValueError(
+                'Somehow the logged in buyer {b} is not configured right '
+                'in Solitude; result: {r}'
+                .format(b=buyer_uuid, r=buyer_results))
+
+        try:
+            resource = api.braintree.mozilla.buyer(buyer['resource_pk'])
+            resource.get_object_or_404()
+        except ObjectDoesNotExist:
+            customer_exists = False
+        else:
+            log.info('found solitude object {o} for buyer {b}'
+                     .format(b=buyer_uuid, o=buyer))
+            customer_exists = True
+
+        if not customer_exists:
+            log.info('creating new braintree customer for {buyer}'
+                     .format(buyer=buyer_uuid))
+            api.braintree.customer.post({'uuid': buyer_uuid})
+        else:
+            log.info('using existing braintree customer tied to buyer {b}'
+                     .format(b=buyer))
```

Comment:
any advantage to having this in a seperate if block, feels like they could be moved into the try / except / else above and compressed a little